src/rust-crypto/PerSessionKeyBackupDownloader.ts

```diff
@@ -0,0 +1,576 @@
+            try {
+                return await olmMachine.getBackupKeys();
+            } catch (e) {
+                return null;
+            }
+        },
+
+        async requestRoomKeyFromBackup(version: string, roomId: string, sessionId: string): Promise<KeyBackupSession> {
+            const path = encodeUri("/room_keys/keys/$roomId/$sessionId", {
+                $roomId: roomId,
+                $sessionId: sessionId,
+            });
+
+            return await http.authedRequest<KeyBackupSession>(Method.Get, path, { version }, undefined, {
+                prefix: ClientPrefix.V3,
+            });
+        },
+
+        async importRoomKeys(keys: IMegolmSessionData[]): Promise<void> {
+            return rustCrypto.importRoomKeys(keys);
+        },
+
+        createBackupDecryptor: (key: RustSdkCryptoJs.BackupDecryptionKey): RustBackupDecryptor => {
+            return new RustBackupDecryptor(key);
+        },
+
+        async requestKeyBackupVersion(): Promise<KeyBackupInfo | null> {
+            return await backupManager.requestKeyBackupVersion();
+        },
+
+        getCryptoEventEmitter(): TypedEventEmitter<RustBackupCryptoEvents, RustBackupCryptoEventMap> {
+            return rustCrypto;
+        },
+    };
+}
+
+/**
+ * Enumerates the different kind of errors that can occurs when downloading and importing a key from backup.
+ */
+export enum KeyDownloadError {
+    /** The backup version in use is out of sync with the server version. */
+    VERSION_MISMATCH = "VERSION_MISMATCH",
+    /** The requested key is not in the backup. */
+    MISSING_DECRYPTION_KEY = "MISSING_DECRYPTION_KEY",
+    /** A network error occurred while trying to get the key. */
+    NETWORK_ERROR = "NETWORK_ERROR",
+    /** The loop as been stopped. */
+    STOPPED = "STOPPED",
+    /** An unknown error occurred while decrypting/importing the key */
+    UNKNOWN_ERROR = "UNKNOWN_ERROR",
+    /** The server is rate limiting us. */
+    RATE_LIMITED = "RATE_LIMITED",
+    /** The backup is not configured correctly, can be that there is no backup, that it's not trusted
+     * , that we don't have the correct key in cache... */
+    CONFIGURATION_ERROR = "CONFIGURATION_ERROR",
+}
+
+/** Helper type for requested session*/
+type SessionInfo = { roomId: string; megolmSessionId: string };
+
+/** Helper type for the result of a key download. */
+type KeyDownloadResult =
+    | { ok: true; value: KeyBackupSession }
+    | { ok: false; error: KeyDownloadError; [key: string]: any };
+
+/** Holds the current backup decryptor and version that should be used. */
+type Configuration = {
+    backupVersion: string;
+    decryptor: RustBackupDecryptor;
+};
+
+/**
+ * Signaling for the Downloader loop.
+ * Not yet used by API, yet useful for testing.
+ */
+export enum KeyDownloaderEvent {
+    DownloadLoopStateUpdate = "download_loop_started",
+    DownLoopStep = "download_loop_step",
+    QueryKeyError = "query_key_error",
+    KeyImported = "key_imported",
```

Comment:
If these events are only used in the unit test, then I'd really like us to look hard at whether we can find other ways to write the test. It feels wrong to have to cover the code in "emit(this)" and "emit(that)" just for the benefit of testing. For example: can we infer the behaviour of the download loop from the calls to the lower-level components?

 Some of them (`DownLoopStep`?) don't even seem to be used at all.

If we keep the events:

* The names don't match the conventions we use elsewhere in the js-sdk.
* Also they need documentation as to what they indicate and what the payloads are.